### PR TITLE
Version 2.0

### PR DESCRIPTION
**Pull request by VladimirPodolian**

### CODE REVIEW

File: dyatel/base/driver_wrapper.py [hunk 1]
```diff
@@ -26,39 +30,40 @@ class DriverWrapper(WebDriver, MobileDriver, PlayDriver):
...
+        return super().__new__(type(f'ShadowDriverWrapper', (cls, ), get_attributes_from_object(cls)))  # noqa
+
+    def __repr__(self):
+        cls = self.__class__
+
+        label = 'desktop'
+        if cls.is_android:
+            label = 'android'
+        elif cls.is_ios:
+            label = 'ios'
```
Comment (Coldshard) [inline dyatel/base/driver_wrapper.py:48]:
```
os = {'android': cls.is_android, 'ios': cls.is_ios, 'desktop': True}
label = next(name for name, state in os.items if state)
```

File: dyatel/base/element.py [hunk 2]
```diff
@@ -210,6 +250,59 @@ def wait_element_hidden_without_error(self, timeout: int = WAIT_EL, silent: bool
...
+        :param: silent: erase log
+        :return: self
+        """
+        if not silent:
+            self.log(f'Wait until "{self.name}" become enabled')
+
+        element = self.element
+        enabled = False
+        start_time = time.time()
+        while time.time() - start_time < timeout and not enabled:
```
Comment (Coldshard) [inline dyatel/base/element.py:267]:
```
from time import time
time()
```

File: dyatel/base/element.py [hunk 3]
```diff
@@ -268,33 +369,46 @@ def assert_screenshot(self, filename: str = '', test_name: str = '', name_suffix
         :param fill_background: fill background with given color or black color by default
         :return: None
         """
+        delay = delay if delay else VisualComparison.default_delay
+        threshold = threshold if threshold else VisualComparison.default_threshold
+        remove = [remove] if type(remove) is not list and remove else remove
```
Comment (Coldshard) [inline dyatel/base/element.py]:
```
        delay = delay or VisualComparison.default_delay
        threshold = threshold or VisualComparison.default_threshold
```

File: dyatel/base/element.py [hunk 4]
```diff
@@ -268,33 +369,46 @@ def assert_screenshot(self, filename: str = '', test_name: str = '', name_suffix
...
+    def _modify_object(self):
+        """
+        Modify current object. Required for Element that placed into functions:
+        - set driver from previous object if previous driver different.
+        - set parent from previous object if previous is Group.
+        """
+        prev_object_manager = PreviousObjectDriver()
+        prev_object_manager.set_driver_from_previous_object_for_element(self, 6)
+        if not self._initialized and self.parent is None:
+            prev_object_manager.set_parent_from_previous_object_for_element(self, 6)
```
Comment (Coldshard) [inline dyatel/base/element.py:414]: Put magic numbers in constants

File: tests/static_tests/functional/test_repr_multiple_driver.py [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+    assert '2_driver' in desktop_info
+    assert 'selenium.webdriver' in desktop_info
+
+
+@pytest.mark.parametrize('driver', mobile_drivers, ids=mobile_ids)
+def test_mobile_and_desktop_page_repr(driver, request):
+    request.getfixturevalue(driver)
+    request.getfixturevalue('mocked_selenium_driver')
+    mobile_driver = request.getfixturevalue(driver)
+    desktop_driver = request.getfixturevalue('mocked_selenium_driver')
```
Comment (Coldshard) [inline tests/static_tests/functional/test_repr_multiple_driver.py]: why we call these functions twice?

File: dyatel/base/driver_wrapper.py (hunk 1 above)
Comment (VladimirPodolian) [reply dyatel/base/driver_wrapper.py:48]: too juicy

File: dyatel/base/element.py (hunk 2 above)
Comment (VladimirPodolian) [reply dyatel/base/element.py:267]: it may be implicit

File: tests/static_tests/functional/test_driver.py [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+    desktop_page = PageObject(driver_wrapper=mocked_selenium_driver)
+
+    assert mobile_page.driver_wrapper == mocked_android_driver
+    assert desktop_page.driver_wrapper == mocked_selenium_driver
+
+    assert mobile_page.section.driver_wrapper == mocked_android_driver
+    assert desktop_page.section.driver_wrapper == mocked_selenium_driver
+
+    assert mobile_page.section.el.driver_wrapper == mocked_android_driver
+    assert desktop_page.section.el.driver_wrapper == mocked_selenium_driver
```
Comment (Coldshard) [inline tests/static_tests/functional/test_driver.py:30]:
```
for driver in [mocked_android_driver, mocked_selenium_driver]:
    for page in [mobile_page, desktop_page]:
        assert page. driver_wrapper == driver
```

File: tests/static_tests/functional/test_child_elements.py [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+def test_child_elements(mocked_selenium_driver):
+    section2 = Section2('.section2', name='section2')
+    section3 = Section3()
+    assert section3.some_element
+    assert section3.some_element._initialized  # noqa
+    assert section3.child_elements
+
+    assert section2.some_element
+    assert section2.some_element._initialized  # noqa
+    assert section2.child_elements
```
Comment (Coldshard) [inline tests/static_tests/functional/test_child_elements.py:36]:
```
for section in [section2, section3]:
    assert section.some_element
    assert section.some_element._initialized  # noqa
    assert section.child_elements
```

File: tests/static_tests/conftest.py [hunk 8]
```diff
@@ -59,15 +57,44 @@ def mocked_selenium_driver():
...
+mobile_drivers = [mocked_ios_driver.__name__, mocked_android_driver.__name__]
+mobile_ids = ['appium ios', 'appium android']
+
+
+desktop_drivers = [mocked_selenium_driver.__name__, mocked_play_driver.__name__]
+desktop_ids = ['selenium', 'playwright']
+
+
+all_drivers = mobile_drivers + desktop_drivers
+all_ids = mobile_ids + desktop_ids
```
Comment (Coldshard) [inline tests/static_tests/conftest.py:96]: Maybe make it a dict like `{id: driver}`?

File: dyatel/visual_comparison.py [hunk 9]
```diff
@@ -242,30 +265,54 @@ def _get_screenshot_name(self, test_function_name: str = '', name_suffix: str =
...
+        mask = numpy.zeros(reference_img.shape, dtype='uint8')
+        filled_after = actual_img.copy()
+
+        for c in contours:
+            area = cv2.contourArea(c)
+            if area > 40:
+                x, y, w, h = cv2.boundingRect(c)
+                cv2.rectangle(reference_img, (x, y), (x + w, y + h), self.diff_color_scheme, 2)
+                cv2.rectangle(actual_img, (x, y), (x + w, y + h), self.diff_color_scheme, 2)
+                cv2.rectangle(diff_box, (x, y), (x + w, y + h), self.diff_color_scheme, 2)
```
Comment (Coldshard) [inline dyatel/visual_comparison.py:310]:
```
for img in [reference_img, actual_img, diff_box]:
    cv2.rectangle(img, (x, y), (x + w, y + h), self.diff_color_scheme, 2)
```

File: dyatel/mixins/selector_synchronizer.py [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+    if locator in all_tags:
+        return locator
+    elif '/' in locator:
+        return f'xpath={locator}'
+    elif '/' not in locator and brackets:
+        return locator
+    elif '.' in locator and not brackets:
+        return locator
+    elif '#' in locator:
+        return locator
```
Comment (Coldshard) [inline dyatel/mixins/selector_synchronizer.py:128]:
```
base_cases = [
    'xpath=' in locator or 'id=' in locator,
    locator in all_tags,
    '/' not in locator and brackets,
    '.' in locator and not brackets,
    '#' in locator
]
if any(base_cases):
    return locator
```

File: dyatel/mixins/selector_synchronizer.py [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+        locator = f'[id="{locator}"]'
+    elif locator_type == By.TAG_NAME:
+        locator_type = By.CSS_SELECTOR
+    elif locator_type == By.CLASS_NAME:
+        locator_type = By.CSS_SELECTOR
+        locator = f".{locator}"
+    elif locator_type == By.NAME:
+        locator_type = By.CSS_SELECTOR
+        locator = f'[name="{locator}"]'
+    return locator, locator_type
```
Comment (Coldshard) [inline dyatel/mixins/selector_synchronizer.py]:
```
    if locator_type == By.ID:
        locator = f'[id="{locator}"]'
    elif locator_type == By.CLASS_NAME:
        locator = f".{locator}"
    elif locator_type == By.NAME:
        locator = f'[name="{locator}"]'
    locator_type = By.CSS_SELECTOR
    return locator, locator_type
```

File: dyatel/mixins/selector_synchronizer.py [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+    elif ':id' in locator:  # Mobile com.android selector
+        return By.ID
+    elif '/' in locator:
+        return By.XPATH
+    elif '/' not in locator and brackets:
+        return By.CSS_SELECTOR
+    elif '.' in locator and not brackets:
+        return By.CSS_SELECTOR
+    elif '#' in locator:
+        return By.CSS_SELECTOR
```
Comment (Coldshard) [inline dyatel/mixins/selector_synchronizer.py:73]:
```
elif ('/' not in locator and brackets) or ('.' in locator and not brackets) or '#' in locator:
    return By.CSS_SELECTOR
```

File: dyatel/mixins/selector_synchronizer.py [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+
+    for tag in locator.split(' '):
+        is_only_tags = is_only_tags and tag in all_tags
```
Comment (Coldshard) [inline dyatel/mixins/selector_synchronizer.py:76]: `is_only_tags = next(tag not in all_tags for tag in locator.split(' '), True)`

**Pull request merged**